File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */



class Solution {
public:
    class Pair {
    public:
        int forwardSlash = -1;
        int backwardSlash = -1;
        int maxZZ = 0;
    };
    
    Pair findZigZag(TreeNode* root){
        if(!root) {
            Pair nodeAns;
            return nodeAns;
        }
        
        Pair left = findZigZag(root->left);
        Pair right = findZigZag(root->right);
        
        Pair nodeAns;
        
        nodeAns.maxZZ = max(max(left.maxZZ, right.maxZZ) ,max(left.backwardSlash, right.forwardSlash)+1);
        nodeAns.forwardSlash = left.backwardSlash +1;
        nodeAns.backwardSlash = right.forwardSlash +1;
        
        return nodeAns;
    }
    
    
    int longestZigZag(TreeNode* root) {
        Pair maxi = findZigZag(root);
        
        return maxi.maxZZ;
    }
};

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int longestZigZag(TreeNode root) {
        Pair p=longestZigZag_(root);
        return p.maxLen;
    }
    
    public Pair longestZigZag_(TreeNode root){
        if(root==null){
            return new Pair();
        }
        
        Pair left=longestZigZag_(root.left);
        Pair right=longestZigZag_(root.right);
        
        Pair ans=new Pair();
        
        ans.maxLen=Math.max(left.maxLen,Math.max(right.maxLen,Math.max(left.backwardSlope,right.forwardSlope)+1));
        ans.forwardSlope=left.backwardSlope+1;
        ans.backwardSlope=right.forwardSlope+1;
        
        return ans;
    }
}

class Pair{
    int forwardSlope=-1;
    int backwardSlope=-1;
    int maxLen=0;
}
